File: Icthus/ReadingViewController.m
//
//  ReadingViewController.m
//  Icthus
//
//  Created by Matthew Lorentz on 8/27/13.
//  Copyright (c) 2013 Matthew Lorentz. All rights reserved.
//

#import "ReadingViewController.h"
#import "ReadingView.h"
#import "BookLocation.h"
#import "AppDelegate.h"
#import "IcthusColorMode.h"
#import "VerseOverlayView.h"
#import "Icthus-Swift.h"

@interface ReadingViewController ()

@property UIBarButtonItem *settingsButton;

@end

@implementation ReadingViewController

@synthesize appDel;
@synthesize book = _book;
@synthesize masterPopover;
@synthesize chapterPickerPopover;
UIColor *tintColor;
CGRect previousFrame;

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
    }
    return self;
}

-(void)awakeFromNib {
    [super awakeFromNib];
    
    self.appDel = (AppDelegate *)[[UIApplication sharedApplication] delegate];
    ColorManager *colorManager = self.appDel.colorManager;
    [self subscribeToColorChangedNotification];

    // Style the nav bar
    self.view.backgroundColor = colorManager.bookBackgroundColor;
    self.navigationController.navigationBar.tintColor = colorManager.tintColor;
    self.navigationController.navigationBar.translucent = colorManager.navBarTranslucency;
    self.navigationController.navigationBar.barTintColor = colorManager.navBarColor;
    UIFont *titleFont;
    if (self.traitCollection.horizontalSizeClass == UIUserInterfaceSizeClassRegular) {
        titleFont = [UIFont fontWithName:@"Avenir-Medium" size:22.0f];
    } else {
        titleFont = [UIFont fontWithName:@"Avenir-Medium" size:20.0f];
    }
    self.navigationController.navigationBar.titleTextAttributes = @{
        NSForegroundColorAttributeName: colorManager.titleTextColor,
        NSFontAttributeName: titleFont,
    };
    
    // Style the Go To button
    if (self.goToButton) {
        self.goToButton.tintColor = colorManager.tintColor;
//        [self.goToButton setTitleTextAttributes:@{
//            NSFontAttributeName: [UIFont fontWithName:@"Bariol-Regular" size:23.0],
//        } forState:UIControlStateNormal];
    }
    
    // Start the NSUserActivity
    if(NSClassFromString(@"NSUserActivity")) {
        self.userActivity = [[NSUserActivity alloc] initWithActivityType:@"com.MattLorentz.Icthus.Reading"];
        self.userActivity.title = @"Reading";
        [self.userActivity becomeCurrent];
        [self setBookToLatest];
    }
}

- (BookLocation *)getLatestLocation {
    self.moc = self.appDel.managedObjectContext;
    NSEntityDescription *entityDescription = [NSEntityDescription entityForName:@"BookLocation" inManagedObjectContext:self.moc];
    NSFetchRequest *request = [[NSFetchRequest alloc] init];
    [request setEntity:entityDescription];
    // Sort by lastModified
    NSSortDescriptor *sortDescriptor = [[NSSortDescriptor alloc] initWithKey:@"lastModified" ascending:NO];
    [request setSortDescriptors:@[sortDescriptor]];
    // Only fetch the most recent location
    [request setFetchLimit:1];
    
    NSError *error;
    NSArray *array = [self.moc executeFetchRequest:request error:&error];
    if (error) {
        NSLog(@"%@", [error localizedDescription]);
    }
    if ([array count]) {
        return [array firstObject];
    } else {
        return nil;
    }
    
}

- (void)setBookToLatest {
    // Warning: This method does not save the current book before changing books.
    self.moc = self.appDel.managedObjectContext;
    // Find the last book that was open and open to it.
    BookLocation *location = [self getLatestLocation];
    if (location) {
        [self setLocation:location];
    } else {
        // Default to Genesis 1:1
        NSFetchRequest *genesisRequest = [NSFetchRequest fetchRequestWithEntityName:@"Book"];
        NSString *translationCode = [[NSUserDefaults standardUserDefaults] objectForKey:@"selectedTranslation"];
        [genesisRequest setPredicate:[NSPredicate predicateWithFormat:@"code == %@ && translation == %@", @"GEN", translationCode]];
        NSError *error;
        NSArray *array = [self.moc executeFetchRequest:genesisRequest error:&error];
        if (error) {
            NSLog(@"%@", [error localizedDescription]);
        } else if ([array count]) {
            Book *genesis = [array firstObject];
            location = [NSEntityDescription insertNewObjectForEntityForName:@"BookLocation" inManagedObjectContext:self.moc];
            [location setBook:genesis chapter:1 verse:1];
            [self.moc save:&error];
            if (error) {
                NSLog(@"%@", [error localizedDescription]);
            }
            
            [self setLocation:location];
        }
    }
}

- (id)initWithBook:(Book *)book {
    self = [super init];
    if (self) {
        self.book = book;
    }
    return self;
}

- (void)splitViewController:(UISplitViewController *)svc willShowViewController:(UIViewController *)aViewController invalidatingBarButtonItem:(UIBarButtonItem *)barButtonItem {
    [self.navigationItem setLeftBarButtonItem:nil animated:YES];
    self.masterPopover = nil;
}

- (void)splitViewController:(UISplitViewController *)svc willHideViewController:(UIViewController *)aViewController withBarButtonItem:(UIBarButtonItem *)barButtonItem forPopoverController:(UIPopoverController *)pc {
    UIImage *settingsIcon = [UIImage imageNamed:@"SettingsIcon"];
    UIBarButtonItem *settingsButton = [[UIBarButtonItem alloc] initWithImage:settingsIcon style:UIBarButtonItemStylePlain target:self.appDel.masterView action:@selector(toggleSettingsPopover)];
    
    UIBarButtonItem *readingListButton = [[UIBarButtonItem alloc] initWithTitle:@"History" style:UIBarButtonItemStylePlain target:self.appDel.masterView action:@selector(toggleReadingListPopover)];

    [self.navigationItem setLeftBarButtonItems:@[settingsButton, readingListButton]];
    self.masterPopover = pc;
}

- (void)setBook:(Book *)newBook {
    _book = newBook;
    [self configureViewWithLocation:[newBook getLocation]];
}

- (void)setLocation:(BookLocation *)location {
    _book = location.book;
    [self configureViewWithLocation:location];
}

- (void)setTranslation:(Translation *)translation {
    if (self.book) {
        NSFetchRequest *request = [NSFetchRequest fetchRequestWithEntityName:@"Book"];
        [request setPredicate:[NSPredicate predicateWithFormat:@"code == %@ && translation == %@", [self.book code], [translation code]]];
        NSError *error;
        NSArray *array = [self.appDel.managedObjectContext executeFetchRequest:request error:&error];
        if (error) {
            NSLog(@"%@", [error localizedDescription]);

        } else {
            self.book = [array firstObject];
            [self configureViewWithLocation:[self.book getLocation]];
        }
    }
}

- (void)configureViewWithLocation:(BookLocation *)location {
    // Update the user interface for the detail item.
    if (self.book) {
        self.navigationItem.title = [self.book shortName];
        if (!self.navigationController) {
            NSLog(@"navigation controller was lost");
        }
        
        [self.readingView setBook:self.book];
        [self.readingView setText:[self.book text]];
        [self.readingView setHorizontalSizeClass:self.traitCollection.horizontalSizeClass];
        [self.readingView buildFrames];
        [self.readingView setCurrentLocation:location];
        [self.readingView addVerseOverlayViewToViewHierarchy];
        NSLog(@"ReadingViewController: Changing book to %@ %@:%@", [self.book shortName], [location chapter], [location verse]);
    }
    
    if (self.masterPopover != nil) {
        [self.masterPopover dismissPopoverAnimated:YES];
    }
    if (self.chapterPickerPopover != nil) {
        [self.chapterPickerPopover dismissPopoverAnimated:YES];
    }
}


- (BOOL)shouldPerformSegueWithIdentifier:(NSString *)identifier sender:(id)sender {
    if ([identifier isEqualToString:@"showChapterPickerPopover"]) {
        if (self.chapterPickerPopover.popoverVisible) {
            [self.chapterPickerPopover dismissPopoverAnimated:YES];
            return NO;
        } else {
            return YES;
        }
    }
    return YES;
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([[segue identifier] isEqualToString:@"showChapterPickerPopover"]) {
        self.chapterPickerPopover = [(UIStoryboardPopoverSegue *)segue popoverController];
    }
    if ([[segue identifier] isEqualToString:@"showHistoryFromLeft"] || [[segue identifier] isEqualToString:@"showSettingsFromLeft"]) {
        [self setModalPresentationStyle:UIModalPresentationOverCurrentContext];
        UIViewController *destination = segue.destinationViewController;
        [destination setModalPresentationStyle:UIModalPresentationOverCurrentContext];
        destination.transitioningDelegate = self;
    }
}

- (id<UIViewControllerAnimatedTransitioning>)animationControllerForPresentedController:(UIViewController *)presented presentingController:(UIViewController *)presenting sourceController:(UIViewController *)source {
    return [[SlideFromLeftPresentationController alloc] init];
}

- (id<UIViewControllerAnimatedTransitioning>)animationControllerForDismissedController:(UIViewController *)dismissed {
     SlideFromLeftPresentationController *animator = [[SlideFromLeftPresentationController alloc] init];
    [animator setPresenting:false];
    return animator;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    [self configureViewWithLocation:[self.book getLocation]];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)viewDidLayoutSubviews {
    if (!CGRectEqualToRect(previousFrame, self.view.frame)) {
        [self.readingView saveCurrentLocation];
        [self.readingView setHorizontalSizeClass:self.traitCollection.horizontalSizeClass];
        [self.readingView redrawText];
        previousFrame = self.view.frame;
        [self setBookToLatest];
    }
    
    // Set up the nav bar
    if (self.settingsButton == nil) {
        UIImage *settingsIcon = [UIImage imageNamed:@"SettingsIcon"];
        self.settingsButton = [[UIBarButtonItem alloc] initWithImage:settingsIcon style:UIBarButtonItemStylePlain target:self action:@selector(settingsButtonTapped)];
    }
    
    // Remove the Settings button
    NSMutableArray<UIBarButtonItem *> *leftBarButtonItems = [[NSMutableArray alloc] initWithArray:self.navigationItem.leftBarButtonItems];
    [leftBarButtonItems removeObject:self.settingsButton];
    
    // Add it back if we have space
    if (self.traitCollection.horizontalSizeClass == UIUserInterfaceSizeClassRegular) {
        [leftBarButtonItems insertObject:self.settingsButton atIndex:0];
    }
    
    [self.navigationItem setLeftBarButtonItems:leftBarButtonItems];
}

- (void)viewWillAppear:(BOOL)animated {

}

- (void)viewWillDisappear:(BOOL)animated {
    // hack to fix a weird bug where self.book would be null on first launch on an 32-bit iPhone using iCloud.
    if (!self.book.managedObjectContext) {
        _book = (Book *)[self.moc objectWithID:self.book.objectID];
    }
    [self.readingView saveCurrentLocation];
}

- (void)viewDidAppear:(BOOL)animated {
    if (![self.readingView.verseOverlayView superview]) {
        [self.readingView addVerseOverlayViewToViewHierarchy];
    }
}

- (void)subscribeToColorChangedNotification {
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(handleColorModeChanged) name:colorModeChangedNotification object:nil];
}

- (void)unsubscribeFromColorChangedNotification {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)handleColorModeChanged {
    ColorManager *colorManager = self.appDel.colorManager;
    self.view.backgroundColor = colorManager.bookBackgroundColor;
    self.navigationController.navigationBar.tintColor = colorManager.tintColor;
    self.navigationController.navigationBar.translucent = colorManager.navBarTranslucency;
    self.navigationController.navigationBar.barTintColor = colorManager.navBarColor;
    self.navigationController.navigationBar.titleTextAttributes = @{
        NSForegroundColorAttributeName: colorManager.titleTextColor,
    };
    if (self.goToButton) {
        self.goToButton.tintColor = colorManager.tintColor;
    }
    [self setBookToLatest];
}

- (IBAction)goToButtonPressed:(id)sender {
    if (self.traitCollection.horizontalSizeClass == UIUserInterfaceSizeClassRegular) {
        [self performSegueWithIdentifier:@"showChapterPickerAsPopover" sender:self];
    } else {
        [self performSegueWithIdentifier:@"showChapterPickerModally" sender:self];
    }
}

- (void)settingsButtonTapped {
    [self performSegueWithIdentifier:@"showSettingsFromLeft" sender:self];
}

- (IBAction)userSwipedFromLeftEdge:(id)sender {
    [self performSegueWithIdentifier:@"showHistoryFromLeft" sender:self];
}

- (void)updateUserActivityState:(NSUserActivity *)activity {
    BasicBookLocation *location = [self.readingView getCurrentLocation];
    activity.userInfo = @{
                          @"bookCode": self.book.code,
                          @"chapter": [NSNumber numberWithInt:location->chapter],
                          @"verse": [NSNumber numberWithInt:location->verse],
                          };
}

- (BOOL)canPerformUnwindSegueAction:(SEL)action fromViewController:(UIViewController *)fromViewController withSender:(id)sender {
    if (action == @selector(unwindToReadingViewController:)) {
        return YES;
    } else {
        return NO;
    }
}

- (IBAction)unwindToReadingViewController:(UIStoryboardSegue *)segue {
}

- (void)dealloc {
    [self unsubscribeFromColorChangedNotification];
    [self.userActivity invalidate];
}


@end

File: Icthus/ReadingViewController.swift
//
//  ReadingViewController.swift
//  Icthus
//
//  Created by Matthew Lorentz on 5/22/15.
//  Copyright (c) 2015 Matthew Lorentz. All rights reserved.
//

import Foundation

class ReadingViewController: UIViewController, UIScrollViewDelegate {
    
    //////////////////////////////////////////////////
    // MARK: Public Properties
    //////////////////////////////////////////////////
    
    var currentBook: Book? {
        get {
            if (_book == nil) {
                _book = getDefaultBook()
            }
            
            return _book
        }
        
        set(newBook) {
            _book = newBook
            if let actualBook = newBook {
                ReadingViewStateMachine.bookChanged(self, actualBook, nil).continuation()
            }
        }
    }
    
    var location: BookLocation? {
        didSet {
            if let actualLocation = location {
                ReadingViewStateMachine.locationChanged(self, actualLocation).continuation()
            }
        }
    }
    
    var translation: Translation? {
        didSet {
            if let actualTranslation = translation {
                ReadingViewStateMachine.translationChanged(self, actualTranslation).continuation()
            }
        }
    }
    
    var scrollView: UIScrollView
    var _location: BookLocation?
    var _book: Book?
    var textViewMetadata: Array<BibleTextViewMetadata> = []
    var textViews: Array<BibleTextView?> = []
    
    //////////////////////////////////////////////////
    // MARK: Private Properties
    //////////////////////////////////////////////////
    
    private var appDel: AppDelegate
    private var moc: NSManagedObjectContext
    private var sizeForMetadata: CGSize?
    private var bookForMetadata: Book?
    private var textViewManager: BibleTextViewManager
    private var lastFrameIndex = 0
    private let numberOfFramesToShow = 15
    
    private var currentFrameIndex: Int {
        get {
            if textViewMetadata.count > 0 {
                var yPos: CGFloat = 0
                var index = 0
                for datum in textViewMetadata {
                    yPos += datum.frame.size.height
                    if yPos >= self.scrollView.contentOffset.y {
                        break
                    }
                    index++
                }
                return index
            } else {
                return 0
            }
        }
    }
    
    //////////////////////////////////////////////////
    // MARK: Initializers
    //////////////////////////////////////////////////
    
    required init(coder aDecoder: NSCoder) {
        appDel = UIApplication.sharedApplication().delegate as! AppDelegate
        moc = appDel.managedObjectContext!
        textViewManager = BibleTextViewManager()
        scrollView = UIScrollView()
        super.init(coder: aDecoder)
    }
    
    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: NSBundle?) {
        appDel = UIApplication.sharedApplication().delegate as! AppDelegate
        moc = appDel.managedObjectContext!
        textViewManager = BibleTextViewManager()
        scrollView = UIScrollView()
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
    }
    
    //////////////////////////////////////////////////
    // MARK: View Lifecycle
    //////////////////////////////////////////////////
    
    override func viewDidLoad() {
        scrollView.frame = CGRect(origin: CGPoint(x: 0, y: 0), size: self.view.frame.size)
        scrollView.delegate = self
        scrollView.scrollsToTop = false
        scrollView.alwaysBounceHorizontal = true
        self.view.addSubview(scrollView)
    }
    
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        // If the frame changes, reload text
        guard let book = currentBook else { return }
        ReadingViewStateMachine.sizeChanged(self, book, view.frame.size).continuation()
    }
    
    override func viewWillTransitionToSize(size: CGSize, withTransitionCoordinator coordinator: UIViewControllerTransitionCoordinator) {
        // If the frame changes, reload text
        guard let book = currentBook else { return }
        ReadingViewStateMachine.sizeChanged(self, book, size).continuation()
    }
    
    
    override func viewWillDisappear(animated: Bool) {
        super.viewWillDisappear(animated)
        saveLocation()
    }
    
    
    //////////////////////////////////////////////////
    // MARK: ScrollView Delegate Functions
    //////////////////////////////////////////////////
    
    func scrollViewDidScroll(scrollView: UIScrollView) {
        if currentFrameIndex != lastFrameIndex {
            addAndRemoveTextViews()
        }
    }
    
    func scrollViewDidEndDecelerating(scrollView: UIScrollView) {
        saveLocation()
    }
    
    func scrollViewDidEndDragging(scrollView: UIScrollView, willDecelerate decelerate: Bool) {
        saveLocation()
    }
    
    //////////////////////////////////////////////////
    // MARK: Managing Location
    //////////////////////////////////////////////////
    
    static func getOffsetForLocation(location: BookLocation, contentSize: CGSize, viewport: CGSize, metadata: [BibleTextViewMetadata]) -> CGPoint {
        // Find the metadatum that contains this location.
        if let metadatum = metadata.filter({$0.containsLocation(location)}).first {
            let textView = BibleTextView(metadata: metadatum, book: location.book)
            let offset = textView.getOffsetForLocation(location)
            if let actualOffset = offset {
                let maxY = contentSize.height - viewport.height
                return CGPoint(x: textView.frame.origin.x, y: min(actualOffset.y, maxY))
            }
        } else {
            print("Warning: Could not find a BibleTextViewMetadata to display location \(location.book.shortName) \(location.chapter):\(location.verse)")
        }
        return CGPoint(x: 0, y: 0)
    }
    
    func saveLocation() {
        guard let book = currentBook else {return}
        let location = ReadingViewController.saveLocationWithOffset(scrollView.contentOffset, metadata: textViewMetadata, textViews: textViews, book: book)
        _location = location
    }
    
    static func saveLocationWithOffset(contentOffset: CGPoint, metadata: [BibleTextViewMetadata], textViews: [BibleTextView?], book: Book) -> BookLocation? {
        // Find the textView currently being viewed
        var position: CGFloat = 0
        var i = 0;
        for i = 0; i < metadata.count; i++ {
            let metadatum = metadata[i]
            if contentOffset.y < position + metadatum.frame.size.height {
                break;
            }
            position += metadatum.frame.size.height
        }
        
        if textViews.count < 1 {
            return nil
        }
        
        let currentTextView = textViews[i]
        if let textView = currentTextView {
            // Get the text position of the first character of the first line that's in view
            let firstLineLocation = contentOffset.y - position
            let optionalRelativeCharPosition = textView.closestPositionToPoint(CGPoint(x: textView.frame.origin.x, y: firstLineLocation))
            guard let relativeCharPosition = optionalRelativeCharPosition else { return nil }
            let relativeCharIndex = textView.offsetFromPosition(textView.beginningOfDocument, toPosition: relativeCharPosition)
            let charIndex = textView.metadata.textRange.location + relativeCharIndex
            
            let location = BibleMarkupParser().saveLocationForCharAtIndex(Int32(charIndex), forText: book.text, andBook: book)
            print("Saved location \(book.shortName) \(location?.chapter):\(location?.verse)")
            return location
        } else {
            print("Warning: location could not be saved because the BibleTextView at the current location is not instantiated.")
        }
        
        return nil
    }
    
    
    //////////////////////////////////////////////////
    // MARK: Presenting Text
    //////////////////////////////////////////////////
    
    func needsNewMetadata() -> Bool {
        if let oldSize = sizeForMetadata where
            CGSizeEqualToSize(self.view.frame.size, oldSize) &&
                currentBook?.objectID == bookForMetadata?.objectID {
                    return false
        } else {
            return true
        }
        
    }
    
    func generatedMetadataWithSize(size: CGSize, book: Book) {
        sizeForMetadata = size
        bookForMetadata = book
    }
    
    func addAndRemoveTextViews() {
        // loop through all text views and instantiate the ones close to the current one and destroy all others
        for i in 0..<textViewMetadata.count {
            var textView: BibleTextView? = textViews[i]
            if textViewWithIndexShouldBeInstantiated(i) {
                if textView == nil, let actualBook = currentBook {
                    textView = BibleTextView(metadata: textViewMetadata[i], book: actualBook)
                    scrollView.addSubview(textView!)
                    textViews[i] = textView
                }
            } else {
                if textView != nil {
                    textView?.removeFromSuperview()
                    textViews[i] = nil
                }
            }
        }
    }
    
    func clearTextViews() {
        textViews.map() { $0?.removeFromSuperview() }
        textViews = Array<BibleTextView>()
    }
    
    //////////////////////////////////////////////////
    // MARK: Helper Functions
    //////////////////////////////////////////////////
    
    func getDefaultBook() -> Book? {
        let genesisRequest = NSFetchRequest(entityName: "Book")
        let translationCode: String = NSUserDefaults.standardUserDefaults().objectForKey("selectedTranslation") as! String
        genesisRequest.predicate = NSPredicate(format: "code == 'GEN' && translation == '\(translationCode)'")
        
        do {
            if let result = try moc.executeFetchRequest(genesisRequest) as? [Book] {
                return result[0]
            }
        } catch {
            print("Error, could not get default book from Core Data.")
        }
        
        return nil
    }
    
    private func textViewWithIndexShouldBeInstantiated(textViewIndex: Int) -> Bool {
        let margin = numberOfFramesToShow / 2
        return abs(currentFrameIndex - textViewIndex) <= margin
    }
}